compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrImplicitCastInserter.kt

```diff
@@ -218,18 +222,28 @@ class Fir2IrImplicitCastInserter(
         }
     }
 
+    private fun FirTypeRef.isNullabilityFlexible(): Boolean {
+        if (hasFlexibleNullability()) {
+            return true
+        }
+        val flexibility = coneTypeSafe<ConeFlexibleType>() ?: return false
+        return flexibility.lowerBound.isMarkedNullable != flexibility.upperBound.isMarkedNullable
+    }
+
     private fun FirTypeRef.acceptsNullValues(): Boolean =
-        isMarkedNullable == true || hasEnhancedNullability()
+        containsNull() || hasEnhancedNullability()
 
     private fun IrExpression.insertImplicitNotNullCastIfNeeded(expression: FirExpression): IrExpression {
         // [TypeOperatorLowering] will retrieve the source (from start offset to end offset) as an assertion message.
         // Avoid type casting if we can't determine the source for some reasons, e.g., implicit `this` receiver.
         if (expression.source == null) {
             return this
         }
+        // Cast type massage 1. Remove @EnhancedNullability
+        // Cast type massage 2. Convert it to a non-null variant (in case of @FlexibleNullability)
```

Comment:
https://github.com/JetBrains/kotlin/blob/master/compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/transformations/InsertImplicitCasts.kt#L392

Two lines there did the same massages.

Reply:
Confirmed